src/Squirrel/UpdateManager.DownloadReleases.cs

```diff
@@ -19,38 +20,54 @@ public DownloadReleasesImpl(string rootAppDirectory)
                 this.rootAppDirectory = rootAppDirectory;
             }
 
-            public async Task DownloadReleases(string updateUrlOrPath, IEnumerable<ReleaseEntry> releasesToDownload, Action<int> progress = null, IFileDownloader urlDownloader = null)
+            public async Task DownloadReleases(string updateUrlOrPath, IEnumerable<ReleaseEntry> releasesToDownload1, Action<int> progress = null, IFileDownloader urlDownloader = null)
             {
                 progress = progress ?? (_ => { });
                 urlDownloader = urlDownloader ?? new FileDownloader();
                 var packagesDirectory = Path.Combine(rootAppDirectory, "packages");
+                var releasesToDownload = new List<ReleaseEntry>(releasesToDownload1);
 
                 double current = 0;
                 double toIncrement = 100.0 / releasesToDownload.Count();
 
                 if (Utility.IsHttpUrl(updateUrlOrPath)) {
                     // From Internet
-                    await releasesToDownload.ForEachAsync(async x => {
-                        var targetFile = Path.Combine(packagesDirectory, x.Filename);
-                        double component = 0;
-                        await downloadRelease(updateUrlOrPath, x, urlDownloader, targetFile, p => {
-                            lock (progress) {
-                                current -= component;
-                                component = toIncrement / 100.0 * p;
-                                progress((int)Math.Round(current += component));
-                            }
-                        });
-                        // With a lot of small updates, and since notifications are only sent every half second
-                        // for each one, we can easily miss half or more of the progress on each download.
-                        // To make sure we eventually get to 100% of the whole process, we need to update
-                        // progress (and especially the total in current) to indicate that this one is complete.
-                        lock (progress)
-                        {
-                            current -= component;
-                            component = toIncrement;
-                            progress((int)Math.Round(current += component));
+                    Exception lastException = null;
+                    for (int i = 0; i < 4; i++) { // try up to this many times to get them all
+                        try {
```

Comment:
github's diff is very unhelpful here. The whole block deleted above is retained unchanged (just more indentation); I added the three lines above and the corresponding catch clause and code following.